tests/unit/s2n_npn_extension_test.c

```diff
@@ -199,5 +199,96 @@ int main(int argc, char **argv)
         }
     }
 
+    /* Tests for the NPN extension on the Encrypted Extensions handshake message */
+    {
+        /* s2n_npn_encrypted_should_send */
+        {
+            /* No connection */
+            EXPECT_FALSE(s2n_npn_encrypted_extension.should_send(NULL));
+
+            /* No chosen application protocol */
+            DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(client_conn);
+            EXPECT_FALSE(s2n_npn_encrypted_extension.should_send(client_conn));
+
+            /* Should send if protocol has been negotiated */
+            uint8_t first_protocol_len = strlen(protocols[0]);
+            EXPECT_MEMCPY_SUCCESS(client_conn->application_protocol, protocols[0], first_protocol_len + 1);
+            EXPECT_TRUE(s2n_npn_encrypted_extension.should_send(client_conn));
+        }
+
+        /* s2n_npn_encrypted_extension_send */
+        {
+            DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(client_conn);
+
+            uint8_t first_protocol_len = strlen(protocols[0]);
+            EXPECT_MEMCPY_SUCCESS(client_conn->application_protocol, protocols[0], first_protocol_len + 1);
+
+            DEFER_CLEANUP(struct s2n_stuffer out = { 0 }, s2n_stuffer_free);
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&out, 0));
+            EXPECT_SUCCESS(s2n_npn_encrypted_extension.send(client_conn, &out));
+
+            uint8_t protocol_len = 0;
+            EXPECT_SUCCESS(s2n_stuffer_read_uint8(&out, &protocol_len));
+
+            uint8_t protocol[UINT8_MAX] = { 0 };
+            EXPECT_SUCCESS(s2n_stuffer_read_bytes(&out, protocol, protocol_len));
+            EXPECT_BYTEARRAY_EQUAL(protocol, protocols[0], protocol_len);
+
+            uint8_t padding_len = 0;
+            EXPECT_SUCCESS(s2n_stuffer_read_uint8(&out, &padding_len));
+            EXPECT_TRUE(padding_len > 0);
+            
+            for(size_t i = 0; i < padding_len; i++) {
+                uint8_t byte = UINT8_MAX;
+                EXPECT_SUCCESS(s2n_stuffer_read_uint8(&out, &byte));
+                EXPECT_EQUAL(byte, 0);
+            }
+            EXPECT_EQUAL(s2n_stuffer_data_available(&out), 0);
+        }
+
+        /* NPN Encrypted Extension recv can read NPN Encrypted Extension send */
```

Comment:
Are there some malformed / failure cases we should check? We try to be pretty defensive with extensions.

Reply:
Added a test case for malformed ext and empty ext.